velox/functions/prestosql/tests/utils/FunctionBaseTest.h

```diff
@@ -155,53 +156,109 @@ class FunctionBaseTest : public testing::Test,
+  ///   }
+  ///
+  ///
+  /// If not specified, input logical types are derived from the C++ type
+  /// provided as input using CppToType. In the example above, the `std::string`
+  /// C++ type will be translated to `VARCHAR` logical type, and `int64_t` to
+  /// `BIGINT`.
+  ///
+  /// To override the input logical type, use:
+  ///
+  ///   std::optional<int64_t> hour(std::optional<int32_t> date) {
+  ///     return evaluateOnce<int64_t>("hour(c0)", DATE(), date);
+  ///   }
+  ///
+  /// Custom types like TIMESTAMP_WITH_TIMEZONE() and HYPERLOGLOG() are also
+  /// supported. For multiple parameters, use a list of logical types:
+  ///
+  ///   return evaluateOnce<std::string>(
+  ///       "hmac_sha1(c0, c1)", {VARBINARY(), VARBINARY()}, arg, key);
+  ///
+  ///
+  /// One can also manually specify additional template parameters, in case
+  /// there are problems with template type deduction:
+  ///
+  ///   return evaluateOnce<int64_t, std::string, int64_t>(
+  ///       "fb_xxhash64(c0, c1)", val, seed);
+  ///
+  template <typename TReturn, typename... TArgs>
+  std::optional<TReturn> evaluateOnce(
       const std::string& expr,
-      const std::optional<Args>&... args) {
-    return evaluateOnce<ReturnType>(
+      const std::initializer_list<TypePtr>& types,
+      std::optional<TArgs>... args) {
+    return evaluateOnce<TReturn>(
         expr,
-        makeRowVector(std::vector<VectorPtr>{
-            makeNullableFlatVector(std::vector{args})...}));
+        makeRowVector(unpackEvaluateParams<std::optional<TArgs>...>(
+            std::vector<TypePtr>{types},
+            std::forward<std::optional<TArgs>>(std::move(args))...)));
   }
 
-  template <typename ReturnType, typename Args>
-  std::optional<ReturnType> evaluateOnce(
+  template <typename TReturn, typename... TArgs>
+  std::optional<TReturn> evaluateOnce(
       const std::string& expr,
-      const std::vector<std::optional<Args>>& args,
-      const std::vector<TypePtr>& types,
-      const std::optional<SelectivityVector>& rows = std::nullopt,
-      const TypePtr& resultType = nullptr) {
-    std::vector<VectorPtr> flatVectors;
-    for (vector_size_t i = 0; i < args.size(); ++i) {
-      flatVectors.emplace_back(makeNullableFlatVector(
-          std::vector<std::optional<Args>>{args[i]}, types[i]));
-    }
-    auto rowVectorPtr = makeRowVector(flatVectors);
-    return evaluateOnce<ReturnType>(expr, rowVectorPtr, rows, resultType);
+      std::optional<TArgs>... args) {
+    return evaluateOnce<TReturn>(
+        expr,
+        makeRowVector(unpackEvaluateParams<std::optional<TArgs>...>(
+            {}, std::forward<std::optional<TArgs>>(std::move(args))...)));
   }
 
-  template <typename ReturnType>
-  std::optional<ReturnType> evaluateOnce(
+  template <typename TReturn, typename... TArgs>
+  std::optional<TReturn> evaluateOnce(
+      const std::string& expr,
+      const TypePtr& type,
+      std::optional<TArgs>... args) {
```

Comment:
Do we support more that one argument here? If so, does 'type' apply only to first argument? It would be nice to clarify.

Reply:
This overload is only here so you can do:

```
return evaluateOnce<int64_t>("hour(c0)", DATE(), date)
```

instead of:

```
return evaluateOnce<int64_t>("hour(c0)", {DATE()}, date)
```

I'll add a small comment to clarify.

Reply:
> Do we support more that one argument here?

Yes, through the first overloaded version of this function (pass the types in the initializer_list)